svcdef/resolvetypes.go

```diff
@@ -0,0 +1,71 @@
+package svcdef
+
+// typeBox is used where we know that either a Message or an Enum will need to
+// be returned, but not both, and not neither. If one field is nil, the other
+// is expected to be a valid pointer.
+type typeBox struct {
+	Message *Message
+	Enum    *Enum
+}
+
+// resolveTypes accepts a pointer to a Svcdef and modifies that Svcdef, and its
+// child structs, in place. After the initial creation of Svcdef, each
+// FieldType will have it's name set to the name of the type that it
+// represents, but it won't have a pointer to the full instance of the type
+// with that name. Type resolution is the process of taking the name of each
+// FieldType, then searching for the enum/message with that name and setting
+// the respective *Message/*Enum property of our FieldType to point to that
+// enum/message.
+func resolveTypes(sd *Svcdef) {
+	tmap := newTypeMap(sd)
+	for _, m := range sd.Messages {
+		for _, f := range m.Fields {
+			setType(f.Type, tmap)
```

Comment:
I believe this to be a bug. IIRC the `f` here is a copy used while ranging over `m.Fields`. This can easily be checked by grabbing the index and comparing their addresses.
https://golang.org/ref/spec#For_statements
```
Range expression                          1st value          2nd value

array or slice  a  [n]E, *[n]E, or []E    index    i  int    a[i]       E
string          s  string type            index    i  int    see below  rune
map             m  map[K]V                key      k  K      m[k]       V
channel         c  chan E, <-chan E       element  e  E
```
see second value `a[i]` vs `&a[i]`